chips/nrf52/src/acomp.rs

```diff
@@ -217,9 +217,9 @@ pub struct Comparator<'a> {
 }
 
 impl<'a> Comparator<'a> {
-    const fn new(registers: StaticRef<CompRegisters>) -> Self {
+    pub const fn new() -> Self {
```

Comment:
How are such `new()` functions not marked as `unsafe`? One can instantiate them twice with safe code, which then aliases the registers, violating Rust's fundamental memory safety rules. On the contrary, the previous code only exposed `static mut` variables as public code, and `static mut` required `unsafe` to use.

I think we should come up with a better API.

Reply:
I spoke with @alevy about this, and I do not believe it violates Rust's memory safety rules. `StaticRef` provides access to the underlying registers by wrapping the underlying registers in an `UnsafeCell`, but it is not memory unsafe for multiple references to a `StaticRef` to exist. Notably, this design does expose a footgun in that it is possible to instantiate a peripheral twice, when only one can be hooked up to receive interrupts. But that is actually occasionally useful (for the UART in the panic handler, for example). We could have each new function require a capability to lessen the impact of this footgun, but I am not sure if that is necessary. 

Note that within a given module it was already possible to alias the registers without unsafe by simply using the `const BASE_REGISTERS` multiple times, which did not require unsafe. In fact, there are already instances of code that does so in Tock drivers today.

That being said, I would be happy to hear suggestions for an alternate API.

Reply:
> `StaticRef` provides access to the underlying registers by wrapping the underlying registers in an `UnsafeCell`.

I don't see any `UnsafeCell` https://github.com/tock/tock/blob/master/kernel/src/common/static_ref.rs. But after giving it a second look, the `StaticRef` API indeed looks fine to me.
- There is an `unsafe` function to create a `StaticRef`, the prerequisite being a non-null pointer with `'static` lifetime to a `T` object that is otherwise not aliased in the program. I don't see these prerequisites violated in the peripherals we instanciate.
- Cloning is safe. No issue with that, cloning a pointer is in itself safe.
- Dereferencing to a `&'static T`. This is safe because the pointer passed to `new` has `'static` lifetime, and indeed points to a `T` (in particular, is aligned correctly). It's also fine to dereference clones of a `StaticRef`, as we obtain *immutable* references.

The `UnsafeCell` part is from https://github.com/tock/tock/blob/master/libraries/tock-register-interface/src/registers.rs. This also looks fine to me, w.r.t. the `UnsafeCell` rules (https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html).

- The register API doesn't expose any `&T` or `&mut T` from the `UnsafeCell` in register. Simple `set`/`get` API.
- No multi-threading in Tock.
- I wonder however how that plays with re-entrancy. In normal operation the Tock scheduler doesn't call code re-entrantly (just queues pending interrupts for the main loop to process).
- The panic handler path should be fine, but is not 100% clear to me. What if a panic happens from within the UART code, and that we re-entrantly call UART code again?

> Note that within a given module it was already possible to alias the registers without unsafe by simply using the `const BASE_REGISTERS` multiple times, which did not require unsafe. In fact, there are already instances of code that does so in Tock drivers today.

Good point.

So I think I don't have much concerns left here. In short, it's possible to instantiate a set of registers twice, and have the two instances make unrelated writes to the underlying registers (but not *concurrently* in the multi-threaded sense, as Tock is single-threaded), which would certainly violate the logic of the underlying peripheral, but wouldn't trigger any UB at the Rust language level (so there is no risk that compiler optimizes away or things like that).
I think it's fine, and best to keep `unsafe` outside of that (I'm all for reducing the amount of `unsafe` code to the minimum).

But I still want to note that depending on what the peripheral does of it, logic bugs where a peripheral's registers are put in an inconsistent state (due to multiple instantiations of the same registers) may still trigger very bad stuff (though I don't have a specific example in mind). So it would be nice to have a way to prevent instantiating registers twice (except for those required in panic handlers), to avoid logic bugs, but that's out of this pull request's scope.

Reply:
Thanks for looking into this some more!  Agreed about the potential for logic bugs, but I don't think its that much worse than before where there was widespread use of the globals that allowed for the same bugs, just requiring an `unsafe`.

Reply:
There are some good use cases where it is useful to have multiple instances of a set of registers.

For example with a pin controller. Let's say there is a pin controller that can be used to allow either I2C or SPI. You could pass the pin controller registers to both the I2C and the SPI devices and then have the I2C/SPI device configure the pins before it starts. Obviously you would then need to make sure only 1 is operating at a time (maybe with a virtual_bus capsule).

The same sort of thing would apply to power management, where each device can enable and disable itself as it is started/stopped. For the Apollo3 for example it would be great to only enable devices when accessed, instead of just enabling everything at boot. Stopping devices is another problem that is much more complex.

Reply:
This can be done by making the `const XXX: StaticRef<Registers>` --> `pub const XXX: StaticRef<Registers>` instead. But I don't think we want to take that approach universally, as it will often only make sense when there is some virtualization ensuring concurrent accesses do not lead to unexpected behavior.

For power management the state-of-the-art in Tock previously was anything that wanted to manage power accessing some global `PowerManager` struct. This change at least requires all devices which want to use the struct to take a reference to the power manager when they are instantiated, so it is clearer which peripherals will interact with it.